### PR TITLE
Added tixi3 3.3.0 conan package

### PR DESCRIPTION
**Pull request by rainman110**
Specify library name and version:  **tixi3/3.3.0**

Tixi3 is a xml library based on libxml2 with a very simple API and is used by multiple open source software packages, in particular used by the German Aerospace Center.

Website: https://github.com/DLR-SC/tixi

I was the main author of the library for the last 10 years and I'd like to share the conan package. 

Tixi3 is not yet used by other packages in the conan center index, but we plan to publish more packages that use tixi3.

_Note: This is my first conan package, so I might need more explanations than usually if changes are required._

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by rainman110** [comment]
Hey guys. How can we proceed here? My access to the repo has not yet been approved in #4.

**Comment by uilianries** [comment]
@rainman110 The EAP process is execute 3 times every week. You need to wait. Thank you for your contribution.

File: recipes/tixi3/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        files.rmdir(self, os.path.join(self.package_folder, "share"))
+
+    def package_info(self):
+        self.cpp_info.includedirs = ['include/tixi3']
+        self.cpp_info.libs = ['tixi3']
+
+        if self.settings.os == "Windows":
+            self.cpp_info.system_libs = ['Shlwapi']
+
+        self.cpp_info.set_property("cmake_target_aliases", ["tixi3"])
```
Comment (uilianries) [inline recipes/tixi3/all/conanfile.py:86]:
```suggestion
        self.cpp_info.set_property("cmake_file_name", "tixi3")
        self.cpp_info.set_property("cmake_target_name", "tixi3")
```
Provide only official targets.

File: recipes/tixi3/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+    }
+
+    source_subfolder = "_src_"
+    generators = ["CMakeDeps", "CMakeToolchain"]
```
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:28]: Sorry, but I don't really get, what the alternative would be. I removed the generators line and added

```python
    def layout(self):
        cmake_layout(self, generator=["CMakeDeps", "CMakeToolchain"])
```

which errors in

```
ERROR: tixi3/3.3.0: Error in build() method, line 64
        cmake = CMake(self)
        FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\sigg_ma\\.conan\\data\\tixi3\\3.3.0\\_\\_\\build\\f01f9d7e45f29c0accab403ac1e95fd7266be4c4\\build\\generators\\CMakePresets.json'
```

File: recipes/tixi3/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
+        # tixi is a c library
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
```
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:45]: I was actually developing this with MSVC 2017 and MSVC 2019, so no problems here with those upper lines.

Hence, I'd better stick with the simpler code as
 - Many other recipes have the same code
 - It's less ugly

File: recipes/tixi3/all/test_package/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if can_run(self):
+            bin_path = os.path.join(self.cpp.build.bindirs[0], "tixi3_conan_test")
+            self.run(bin_path, run_environment=True)
```
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/test_package/conanfile.py:23]: Nice one!

File: recipes/tixi3/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+        # copy the LICENSE file
+        self.copy("LICENSE", dst="licenses", src=self.source_subfolder)
+
+        # remove share directory, which only contains documentation,
+        # expamples...
+        files.rmdir(self, os.path.join(self.package_folder, "share"))
+
+    def package_info(self):
+        self.cpp_info.includedirs = ['include/tixi3']
```
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py]: Done

File: recipes/tixi3/all/conanfile.py (hunk 1 above)
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:86]: Thanks for the hint. Actually, the only official target is tixi3, so this was indeed very helpful.

File: recipes/tixi3/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            self.copy(patch["patch_file"])
+
+    def source(self):
+        files.get(self, **self.conan_data["sources"][self.version],
+                  strip_root=True, destination=self.source_subfolder)
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            files.patch(self,
+                patch_file=patch["patch_file"],
+                base_path=self.source_subfolder)
```
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:59]: Done in eaf13b75ee9062a3d5f64b284040507e61d507da

File: recipes/tixi3/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        if self.options.shared:
+            del self.options.fPIC
+
+        # tixi is a c library
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def export_sources(self):
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            self.copy(patch["patch_file"])
```
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py]: Done in eaf13b75ee9062a3d5f64b284040507e61d507da

**Comment by rainman110** [comment]
@uilianries Thanks for the very helpful in-depth code review. I learned quite a bit and addressed most of your comments.

Regarding the cmake_layout changes, I don't know where to start. Do you have a good recipe in mind that I could copy?

From what I've seen, the CMake configuration uses by default an in-source build, which is not supported by tixi. Does the cmake_layout support also out-of-source builds?

File: recipes/tixi3/all/conanfile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:28]: Do you mean something like:

```python
    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()
        deps = CMakeDeps(self)
        deps.generate()
```

???

[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py]: Done in 2f0a4aaa222fbac8a5a73763ba561b4bb57ee2dc

File: recipes/tixi3/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+    }
+
+    source_subfolder = "_src_"
```
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py]: Done in 1cb618aba10bd9cd08f9410e4762f271dd1242c8

**Comment by rainman110** [comment]
@uilianries One remaining question. How can i make the test_v1_package work with the tixi3 target? 

It right now only work with the target `` tixi3::tixi3`` , but I'd like to use the official ``tixi3`` target. Is there any way to change this?

File: recipes/tixi3/all/conanfile.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:51]: Just a note to myself. The conan 2.0 migration guide mentions, that the compiler settings will change and safeguards are required when trying to delete these settings.

[the remark this answers is not in the document]
Comment (rainman110) [reply recipes/tixi3/all/conanfile.py]: Done in  0874a15

**Comment by uilianries** [comment]
@rainman110 Yes, you need to follow this trick: https://github.com/conan-io/conan-center-index/blob/957f561f679f219c0ddd01b7c5d12ba670c27506/recipes/etl/all/conanfile.py#L42

Generate a custom cmake file, which provides an alias, then update cpp_info to say to include that new file.

**Comment by rainman110** [comment]
> Generate a custom cmake file, which provides an alias, then update cpp_info to say to include that new file.

Thanks again @uilianries  :) . As suggested, I created the alias target, which works like a charm. Pushed as 
d9668cb

**Comment by rainman110** [comment]
@uilianries All builds are green. Ready to finish the review process.

File: recipes/tixi3/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+import textwrap
+
+required_conan_version = ">=1.45.0"
+
+
+class Tixi3Conan(ConanFile):
+    name = "tixi3"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "A simple xml interface based on libxml2 and libxslt"
+    topics = ("tixi3", "xml")
```
Comment (uilianries) [inline recipes/tixi3/all/conanfile.py:14]:
```suggestion
    topics = ("xml", "xml2", "xslt")
```
We dont add conan neither the project name as topic.

File: recipes/tixi3/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+    }
+
+    def generate(self):
+        tc = CMakeToolchain(self)
```
Comment (uilianries) [inline recipes/tixi3/all/conanfile.py:29]:
```suggestion
        tc = CMakeToolchain(self)
        tc.variables["TIXI_BUILD_EXAMPLES"] = False
```

File: recipes/tixi3/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        files.get(self, **self.conan_data["sources"][self.version],
+                  strip_root=True, destination=self.source_folder)
+
+    def build(self):
+        files.apply_conandata_patches(self)
+
+        cmake = CMake(self)
+        cmake.configure(variables={
+            "TIXI_BUILD_EXAMPLES": "OFF"
+        })
```
Comment (uilianries) [inline recipes/tixi3/all/conanfile.py:77]:
```suggestion
        cmake.configure()
```
Variables should be declared in the toolchain. More info: https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#the-generate-method

File: recipes/tixi3/all/conandata.yml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+sources:
+  "3.3.0":
+    sha256: "988d79ccd53c815d382cff0c244c0bb8e393986377dfb45385792766adf6f6a9"
+    url: "https://github.com/DLR-SC/tixi/archive/refs/tags/v3.3.0.tar.gz"
+patches:
+  "3.3.0":
+    - patch_file: "patches/link_curl.patch"
+    - patch_file: "patches/disable_tixi_examples.patch"
```
Comment (uilianries) [inline recipes/tixi3/all/conandata.yml:8]:
```suggestion
    - patch_file: "patches/disable_tixi_examples.patch"
      description: "Keep build examples as optional"
      patch_type: "conan"
```
https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches

File: recipes/tixi3/all/conandata.yml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+sources:
+  "3.3.0":
+    sha256: "988d79ccd53c815d382cff0c244c0bb8e393986377dfb45385792766adf6f6a9"
+    url: "https://github.com/DLR-SC/tixi/archive/refs/tags/v3.3.0.tar.gz"
+patches:
+  "3.3.0":
+    - patch_file: "patches/link_curl.patch"
```
Comment (uilianries) [inline recipes/tixi3/all/conandata.yml:7]:
```suggestion
    - patch_file: "patches/link_curl.patch"
      description: "Fix CMake target name for libcurl"
      patch_type: "conan"
```

**Review by uilianries — COMMENTED** [review]
Only small details, but it looks good

File: recipes/tixi3/all/conandata.yml (hunk 12 above)
Comment (rainman110) [reply recipes/tixi3/all/conandata.yml:8]: The conan commit hook errors with your proposed changes:
[HOOK - conan-center.py] pre_export(): ERROR: [CONANDATA.YML FORMAT (KB-H030)] Additional entries ['description'] not allowed in 'patches':'3.3.0' of conandata.yml (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H030)

What should I do?

Comment (uilianries) [reply recipes/tixi3/all/conandata.yml:8]: sorry, it's patch_description

File: recipes/tixi3/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+from conan.tools import files
+from conan import ConanFile
+import os
+import textwrap
+
+required_conan_version = ">=1.45.0"
+
+
+class Tixi3Conan(ConanFile):
+    name = "tixi3"
```
Comment (uilianries) [inline recipes/tixi3/all/conanfile.py:11]: Please, update the recipe name to `tixi` following the project name. Also, you need to update the class name and folder name as well

Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:11]: The name tixi3 is intentional to be able to use both tixi and tixi3 simultaneously. We did the same when creating conda packages for it, there is a tixi and a tixi3 package containing the tixi and tixi3 library.
https://anaconda.org/dlr-sc/repo?type=conda&label=main

Hence, I don't really like to rename this. This is similar to the libxml2 package, which is not called libxml.

Comment (uilianries) [reply recipes/tixi3/all/conanfile.py:11]: @rainman110 I don't get, what do you mean `tixi` project, besides `https://github.com/DLR-SC/tixi`? Are you considering running two different versions at same project?

Comment (rainman110) [reply recipes/tixi3/all/conanfile.py:11]: In our internal bigger build chain, we use both libraries simulateously to be able to update particular "cpacs" xml files. 

We have a geometry processor tigl, whose datamodel has changed a lot with the major version change from 2 to 3. To be able to update the data model and the geometry representation, we developed an update tool using both tigl and tigl3. It will be load tigl library and generate an aircraft geonetry. Tigl requires the tixi library. Then some transformations will be done and the geometry is passed to the tigl3 library, which requires tixi3. This is intentional, as our users need to transition from tigl to tigl3.

Long Story short: we need to be able to link both tixi and tixi3 simultaneously, hence the decision for the tixi3 package name.

From the technical point of view: The official tixi cmake config file called, tixi3-config.cmake with a tixi3 target:
https://github.com/DLR-SC/tixi/blob/master/src/CMakeLists.txt#L86

```cmake
install (FILES
    "${CMAKE_CURRENT_BINARY_DIR}/tixi3-config.cmake"
    "${CMAKE_CURRENT_BINARY_DIR}/tixi3-config-version.cmake"
    DESTINATION ${CONFIG_INSTALL_DIR})
```

It really does not make any sense to use the name "tixi" without a 3 for this package.

Comment (uilianries) [reply recipes/tixi3/all/conanfile.py:11]: @rainman110 Thank you for clarifying. In that case, makes sense keeping tixi3.

**Review by uilianries — APPROVED** [review]
LGTM

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Thanks for taking the time to contribute this really interesting recipe!
